### PR TITLE
95nvmf: add module for NVMe-oF

### PR DESCRIPTION
**Pull request by hreinecke**
Add a module for booting from NVMe-oF devices.

Signed-off-by: Hannes Reinecke <hare@suse.de>

### CODE REVIEW

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    local IFS=:
+    local trtrype="none"
+    local traddr="none"
+    local hosttraddr="none"
+    local trsvcid=4420
+
+    set $fcoe
+    IFS="$OLDIFS"
+
+    set $1
```
Comment (danimo) [inline modules.d/95nvmf/parse-nvmf-boot-connections.sh:28]: Does that work after resetting IFS?

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            warn "host traddr is mandatory for fc"
+            return 1
+        fi
+    elif [ "$trtype" != "rdma" ] && [ "$trtype" != "tcp" ] ; then
+        warn "unsupported transport $trtype"
+        return 1
+    elif [ -z "$trsvcid" ] ; then
+        trsvcid=4420
+    fi
+    echo "--transport=$trtype --traddr=$traddr --host-traddr=$hosttraddr --trsvcid=$trsvdid" >> /etc/nvme/discovery.conf
```
Comment (danimo) [inline modules.d/95nvmf/parse-nvmf-boot-connections.sh:68]: `$trsvdid`... typo?

File: modules.d/95nvmf/module-setup.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+# called by dracut
+check() {
+    require_binaries nvme || return 1
+    [ -f /etc/nvme/hostnqn ] || return 255
+    [ -f /etc/nvme/hostid ] || return 255
+
+    is_nvme_fc() {
+        local _dev=$1
+	local traddr
```
[the remark this answers is not in the document]
Comment (hreinecke) [reply modules.d/95nvmf/module-setup.sh:11]: Sure.

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+# nvmf.discover=fc:auto
+#
+# Note: FC does autodiscovery, so typically there is no need to
+# specify any discover parameters for FC.
+#
+
+parse_nvmf_discover() {
+    local OLDIFS="$IFS"
+    local IFS=:
+    local trtrype="none"
```
[the remark this answers is not in the document]
Comment (hreinecke) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh:20]: Right.

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        3)
+            trtype=$1
+            traddr=$2
+            hosttraddr=$3
+            ;;
+        4)
+            trtype=$1
+            traddr=$2
+            hosttraddr=$3
+            trscvid=$4
```
[the remark this answers is not in the document]
Comment (hreinecke) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh:44]: Hmm? I'm not seeing it ...

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            warn "Invalid arguments for nvmf=$nvmf"
+            return 1
+            ;;
+    esac
+    if [ -z "$traddr" ] ; then
+        warn "traddr is mandatory for $trtype"
+        return 1;
+    fi
+    [ -z "hosttraddr" ] && hosttraddr="none"
+    [ -z "trsvcid" ] && trsvcid="none"
```
[the remark this answers is not in the document]
Comment (hreinecke) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh:56]: Ah. Of course.

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            hosttraddr=$3
+            ;;
+        4)
+            trtype=$1
+            traddr=$2
+            hosttraddr=$3
+            trscvid=$4
+            ;;
+        *)
+            warn "Invalid arguments for nvmf=$nvmf"
```
[the remark this answers is not in the document]
Comment (hreinecke) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh:47]: Should read 'nvmf.discover'. I'll fix it up.

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+# specify any discover parameters for FC.
+#
+
+parse_nvmf_discover() {
+    local OLDIFS="$IFS"
+    local IFS=:
+    local trtrype="none"
+    local traddr="none"
+    local hosttraddr="none"
+    local trsvcid=4420
```
[the remark this answers is not in the document]
Comment (hreinecke) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh:23]: Is that a requirement? If so, sure.

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+    set $fcoe
```
[the remark this answers is not in the document]
Comment (hreinecke) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh:25]: Indeed. Killed.

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh (hunk 2 above)
Comment (hreinecke) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh:68]: Indeed.

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (danimo) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh:42]: trscvid vs trsvcid. (extremely hard to spot. Would have missed it without shellcheck.)

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (danimo) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh]: All boot time scripts should be bashism-free. Some aren't (esp. network). At least we should not lie (and the script specifies `#!/bin/sh`), but here it seems trivial enough to go bashism-free without much hassle :)

[the remark this answers is not in the document]
Comment (hreinecke) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh]: Ok. Will be fixing for the next version.

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh (hunk 1 above)
Comment (hreinecke) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh:25]: Indeed, it doesn't. Fixed in the next version.

File: modules.d/95nvmf/parse-nvmf-boot-connections.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (hreinecke) [reply modules.d/95nvmf/parse-nvmf-boot-connections.sh:42]: Fixed now.

**Comment by hreinecke** [comment]
I've now created a new pull request (#844) which contains all reviews.

**Pull request closed without merging**